src/trace/implementations/chunker.rs

```diff
@@ -0,0 +1,325 @@
+                let mut drain = vec.drain(..).peekable();
+                while drain.peek().is_some() {
+                    self.pending.extend((&mut drain).take(self.pending.capacity() - self.pending.len()));
+                    form_chunk(self);
+                }
+            }
+        }
+    }
+
+    fn extract(&mut self) -> Option<Self::Output> {
+        self.ready.pop()
+    }
+
+    fn finish(&mut self) -> Option<Self::Output> {
+        consolidate_updates(&mut self.pending);
+        while !self.pending.is_empty() {
+            let mut chunk = TimelyStack::with_capacity(self.chunk_capacity());
+            for item in self.pending.drain(..std::cmp::min(self.pending.len(), chunk.capacity())) {
+                chunk.copy(&item);
+            }
+            self.ready.push(chunk);
+        }
+        self.ready.pop()
+    }
+}
+
+/// Chunk a stream of vectors into chains of vectors.
+pub struct ContainerChunker<I, O, Sorter, Consolidator>
+where
+    I: Container,
+    for<'a> O: SizableContainer + PushInto<I::ItemRef<'a>>,
+    Sorter: ContainerSorter<O>,
+    Consolidator: ConsolidateContainer<O> + ?Sized,
+{
+    pending: O,
+    empty: O,
+    ready: Vec<O>,
+    sorter: Sorter,
+    _marker: PhantomData<(I, Consolidator)>,
+}
+
+impl<I, O, Sorter, Consolidator> Default for ContainerChunker<I, O, Sorter, Consolidator>
+where
+    I: Container,
+    for<'a> O: SizableContainer + PushInto<I::ItemRef<'a>>,
+    Sorter: ContainerSorter<O> + Default,
+    Consolidator: ConsolidateContainer<O> + ?Sized,
+{
+    fn default() -> Self {
+        Self {
+            pending: O::default(),
+            empty: O::default(),
+            ready: Vec::default(),
+            sorter: Sorter::default(),
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<I, O, Sorter, Consolidator> Chunker for ContainerChunker<I, O, Sorter, Consolidator>
+where
+    I: Container,
+    for<'a> O: SizableContainer + PushInto<I::ItemRef<'a>>,
+    Sorter: ContainerSorter<O>,
+    Consolidator: ConsolidateContainer<O>,
+{
+    type Input = I;
+    type Output = O;
+
+    fn push_container(&mut self, container: RefOrMut<Self::Input>) {
+        if self.pending.capacity() < O::preferred_capacity() {
+            self.pending.reserve(O::preferred_capacity() - self.pending.len());
+        }
+        // TODO: This uses `IterRef`, which isn't optimal for containers that can move.
+        for item in container.iter() {
+            self.pending.push(item);
+            if self.pending.len() == self.pending.capacity() {
+                self.sorter.sort(&mut self.pending);
+                Consolidator::consolidate_container(&mut self.pending, &mut self.empty);
+                std::mem::swap(&mut self.pending, &mut self.empty);
+                self.empty.clear();
+                if self.pending.len() > self.pending.capacity() / 2 {
+                    self.ready.push(std::mem::take(&mut self.pending));
+                }
```

Comment:
Yes, that's true. The reason for this is that we cannot extract parts of `pending` separately, we have to read it all instead of leaving unfinished chunks behind. Maybe we can think of a better approach, for example keeping pending containers of double the desired capacity around, and once it's more than half full after consolidation, ship one or two chunks, and copy the remaining items into the other pending container, and swap them.

Reply:
I think at least we should drop a comment in here about the decision, because I suspect that the opposite choice is *probably* want we want: spend more CPU copying data (the un-drained part) so that we pack data more densely in the scarce backing resources.